### DISCUSSION TITLE
bean과 enum 매핑

### DISCUSSION THREAD

**Discussion by Dae-Hwa** [Sc0up/scoup-backend#93] [Refactor]
https://github.com/Sc0up/scoup-backend/pull/84#discussion_r693728215

**Comment by janeljs** [comment]
아래와 같이 enum type을 반환하는 메서드를 추가한 뒤 
```java
@Component
public class Bean1 implements EnumMappedBean {
    @Override
    public SomeEnum getSomeEnum() {
        return SomeEnum.ENUM1;
    }
}
```
Config 클래스에 enum과 bean을 맵핑한 빈을 등록하면
```java
@Configuration
public class AppConfig {
    @Bean
    public Map<SomeEnum, EnumMappedBean> getBeansMappedByEnum(@NonNull Collection<EnumMappedBean> enumBeans) {
        return enumBeans.stream()
                .collect(toMap(EnumMappedBean::getSomeEnum, Function.identity()));
    }
}
```
서비스에서 enum type을 키로 사용하는 맵을 주입받을 수 있는 것 같습니다. 
```java
@Service
public class SomeOtherBean {

    private Map<SomeEnum, EnumMappedBean> beansMappedByEnum;

    @Autowired
    public SomeOtherBean(@NonNull Map<SomeEnum, EnumMappedBean> beansMappedByEnum) {
        this.beansMappedByEnum = beansMappedByEnum;
    }
}
```

참고: https://stackoverflow.com/questions/57899018/how-to-injectautowired-beans-to-map-with-enum-as-map-key-in-spring

**Reply by janeljs** [reply]
그런데 `OAuth oAuth = oAuthMap.get(type.name().toLowerCase());` 이 코드 한 줄 없애기 위해 저 코드들을 추가하는게 맞는지는 고민이 되는 거 같아요.
혹시 더 좋은 방법 찾으시면 알려주세요ㅎㅎ

**Reply by Dae-Hwa** [reply]
고생많으셨어요👍
바꾸고자했던건 단순히 저 코드를 없앤다기 보다는 enum의 쓰임새 때문이었습니다.

1. enum이 있으면 최대한 해당 enum을 활용하는게 좋다 생각합니다. `type.name().toLowerCase()`도 사실 제 생각엔 case sensitive해야 맞다고 생각하긴 하는데, 어찌됐건 이런 논란 없이 깔끔하게 정리될 수 있는게 가장 장점이라 생각합니다. key 자체를 원시값이 아닌 타입으로 사용할 수 있으니 애초에 잘못된 값인지 체크가 필요없어지겠죠 요는 enum이 있다면 저희 구조 안에 들어온 이상 enum을 반드시 사용하도록 하는게 좋지 않나 하는 생각입니다.
2. 현재 코드에서 어쩔 수 없다 생각하여 리뷰에 적지는 않았지만 `@Component("github")` 와 같은 컴포넌트 네임은 혼동을 야기할 수 있을 것 같아요. 해당 기능에 맞게 `@Component("gitHubOAuth")` 와 같이 지어주는게 일관성이 있다고 생각합니다.
    > https://docs.spring.io/spring-framework/docs/5.2.3.RELEASE/spring-framework-reference/core.html#beans-beanname

**Reply by janeljs** [reply]
@Dae-Hwa 프레디 말씀을 들어보니까 코드 길이만 고려할게 아니라 enum 타입을 사용하여 맵핑하는게 맞는거 같아요ㅎㅎ 리팩토링하여 pr 보내겠습니다
